kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/Metadata.kt

```diff
@@ -0,0 +1,350 @@
+ * @param [getTypeParameter] a function that returns the type parameter for the given index. **Only
+ *     called if [TypeNameKmTypeVisitor.visitTypeParameter] is called**
+ * @param useTypeAlias indicates whether or not to use type aliases or resolve their underlying
+ *     types
+ */
+internal class TypeNameKmTypeVisitor(flags: Flags,
+    private val getTypeParameter: ((index: Int) -> TypeName)? = null,
+    private val useTypeAlias: Boolean = false,
+    private val receiver: (TypeName) -> Unit) : KmTypeVisitor() {
+
+  private val nullable = Flag.Type.IS_NULLABLE(flags)
+  private var className: String? = null
+  private var typeAliasName: String? = null
+  private var typeAliasType: TypeName? = null
+  private var flexibleTypeUpperBound: TypeName? = null
+  private var outerType: TypeName? = null
+  private var typeParameter: TypeName? = null
+  private val argumentList = mutableListOf<TypeName>()
+
+  override fun visitAbbreviatedType(flags: Flags): KmTypeVisitor? {
+    if (!useTypeAlias) {
+      return null
+    }
+    return TypeNameKmTypeVisitor(flags, getTypeParameter, useTypeAlias) {
+      typeAliasType = it
+    }
+  }
+
+  override fun visitArgument(flags: Flags, variance: KmVariance): KmTypeVisitor? {
+    return TypeNameKmTypeVisitor(flags, getTypeParameter, useTypeAlias) {
+      argumentList.add(
+          when (variance) {
+            KmVariance.IN -> WildcardTypeName.supertypeOf(it)
+            KmVariance.OUT -> {
+              if (it == ANY) {
+                // This becomes a *, which we actually don't want here.
+                // List<Any> works with List<*>, but List<*> doesn't work with List<Any>
+                it
+              } else {
+                WildcardTypeName.subtypeOf(it)
+              }
+            }
+            KmVariance.INVARIANT -> it
+          }
+      )
+    }
+  }
+
+  override fun visitClass(name: kotlinx.metadata.ClassName) {
+    className = name
+  }
+
+  override fun visitFlexibleTypeUpperBound(flags: Flags,
+      typeFlexibilityId: String?): KmTypeVisitor? {
+    return TypeNameKmTypeVisitor(flags, getTypeParameter, useTypeAlias) {
+      flexibleTypeUpperBound = WildcardTypeName.subtypeOf(it)
+    }
+  }
+
+  override fun visitOuterType(flags: Flags): KmTypeVisitor? {
+    return TypeNameKmTypeVisitor(flags, getTypeParameter, useTypeAlias) {
+      outerType = WildcardTypeName.supertypeOf(it)
+    }
+  }
+
+  override fun visitStarProjection() {
+    argumentList.add(WildcardTypeName.STAR)
+  }
+
+  override fun visitTypeAlias(name: kotlinx.metadata.ClassName) {
+    if (!useTypeAlias) {
+      return
+    }
+    typeAliasName = name
+  }
+
+  override fun visitTypeParameter(id: Int) {
+    typeParameter = getTypeParameter?.invoke(id) ?: throw IllegalStateException(
+        "Visiting TypeParameter when there are no type parameters!")
+  }
+
+  override fun visitEnd() {
+    var finalType = flexibleTypeUpperBound ?: outerType ?: typeParameter
```

Comment:
I find this surprising . . . I assume it’ll always be exactly one of these and not multiple?

Reply:
correct. This is probably weird logic to express that, but basically it's "figure out which one of these is present and use that"